[ 0.9.3-nightly-be792b9] Querying data node causes panic

```
[http] 2015/08/06 17:06:37 ::1 - - [06/Aug/2015:17:06:37 +0800] GET /query?q=SELECT+max+FROM+%22Requests%22+&db=throne HTTP/1.1 200 210 http://localhost:8083/ Mozilla/5.0 (Macintosh; Intel Mac OS X 10.10; rv:39.0) Gecko/20100101 Firefox/39.0 712beabc-3c1a-11e5-80ca-000000000000 1.365744ms
panic: interface conversion: interface is float64, not map[string]interface {}

goroutine 408 [running]:
github.com/influxdb/influxdb/tsdb.(*limitedRowWriter).processValues(0xc208504000, 0xc2083ba000, 0xa, 0x2710, 0x0)
        /private/tmp/influxdb20150724-51807-r65xyy/influxdb-0.9.2/src/github.com/influxdb/influxdb/tsdb/engine.go:758 +0xc63
github.com/influxdb/influxdb/tsdb.(*limitedRowWriter).Flush(0xc208504000)
        /private/tmp/influxdb20150724-51807-r65xyy/influxdb-0.9.2/src/github.com/influxdb/influxdb/tsdb/engine.go:693 +0xba
github.com/influxdb/influxdb/tsdb.(*RawExecutor).execute(0xc208560330, 0xc20806a9c0)
        /private/tmp/influxdb20150724-51807-r65xyy/influxdb-0.9.2/src/github.com/influxdb/influxdb/tsdb/engine.go:149 +0x2d7
created by github.com/influxdb/influxdb/tsdb.(*RawExecutor).Execute
        /private/tmp/influxdb20150724-51807-r65xyy/influxdb-0.9.2/src/github.com/influxdb/influxdb/tsdb/engine.go:93 +0x5a
```

I get the same panic when I query a data node in a cluster in version 0.9.3-nightly-be792b9, commit be792b926a86bad31b8e3c9e0d997a6ede31dc70

Steps to repro in AWS:

1. Launch 4 AWS instances and Install Influxdb on each node.
2. On each node update set `hostname = "<Instances Private DNS>"`  in the meta section of `/etc/opt/influxdb/influxdb.conf`.
3. Start InfluxDB on one node.
4. On another node in `/etc/init.d/influxdb` set `INFLUXD_OPTS="-join addr_1:port_1"`.
5. Start InfluxDB on that node.
6. On another node in `/etc/init.d/influxdb` set `INFLUXD_OPTS="-join addr_1:port_1,addr_2:port_2"`.
7. Start InfluxDB on that node.
8. On another node in `/etc/init.d/influxdb` set `INFLUXD_OPTS="-join addr_1:port_1,addr_2:port_2"`.
9. Start InfluxDB on that node.
10. Write one point to each of the nodes.
11. Query the 3 raft nodes (should see all 4 points).
12. Query the 1 data node. (should cause panic).

```
panic: interface conversion: interface is float64, not map[string]interface {}

goroutine 72 [running]:
github.com/influxdb/influxdb/tsdb.(*limitedRowWriter).processValues(0xc2080ea000, 0xc2081ae000, 0x1, 0x2710, 0x0)
	/tmp/tmp.IrAPUmtyJ8/src/github.com/influxdb/influxdb/tsdb/executor.go:734 +0x10bf
github.com/influxdb/influxdb/tsdb.(*limitedRowWriter).Flush(0xc2080ea000)
	/tmp/tmp.IrAPUmtyJ8/src/github.com/influxdb/influxdb/tsdb/executor.go:660 +0xd6
github.com/influxdb/influxdb/tsdb.(*Executor).executeRaw(0xc208180090, 0xc2080801e0)
	/tmp/tmp.IrAPUmtyJ8/src/github.com/influxdb/influxdb/tsdb/executor.go:239 +0x1211
created by github.com/influxdb/influxdb/tsdb.(*Executor).Execute
	/tmp/tmp.IrAPUmtyJ8/src/github.com/influxdb/influxdb/tsdb/executor.go:88 +0x9a
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -15,6 +15,7 @@ With this release InfluxDB is moving to Go 1.5.
 - [#3884](https://github.com/influxdb/influxdb/pull/3884): Fix two panics in WAL that can happen at server startup
 - [#3868](https://github.com/influxdb/influxdb/pull/3868): Add shell option to start the daemon on CentOS. Thanks @SwannCroiset.
 - [#3886](https://github.com/influxdb/influxdb/pull/3886): Prevent write timeouts due to lock contention in WAL
+- [#3574](https://github.com/influxdb/influxdb/issues/3574): Querying data node causes panic
 
 ## v0.9.3 [2015-08-26]
 

--- a/cluster/shard_mapper.go
+++ b/cluster/shard_mapper.go
@@ -2,7 +2,6 @@ package cluster
 
 import (
 	"fmt"
-	"io"
 	"math/rand"
 	"net"
 	"time"
@@ -10,7 +9,6 @@ import (
 	"github.com/influxdb/influxdb/influxql"
 	"github.com/influxdb/influxdb/meta"
 	"github.com/influxdb/influxdb/tsdb"
-	"gopkg.in/fatih/pool.v2"
 )
 
 // ShardMapper is responsible for providing mappers for requested shards. It is
@@ -55,32 +53,26 @@ func (s *ShardMapper) CreateMapper(sh meta.ShardInfo, stmt influxql.Statement, c
 		}
 		conn.SetDeadline(time.Now().Add(s.timeout))
 
-		m.SetRemote(NewRemoteMapper(conn.(*pool.PoolConn), sh.ID, stmt, chunkSize))
+		m.SetRemote(NewRemoteMapper(conn, sh.ID, stmt, chunkSize))
 	}
 
 	return m, nil
 }
 
 func (s *ShardMapper) dial(nodeID uint64) (net.Conn, error) {
-	// If we don't have a connection pool for that addr yet, create one
-	_, ok := s.pool.getPool(nodeID)
-	if !ok {
-		factory := &connFactory{nodeID: nodeID, clientPool: s.pool, timeout: s.timeout}
-		factory.metaStore = s.MetaStore
-
-		p, err := pool.NewChannelPool(1, 3, factory.dial)
-		if err != nil {
-			return nil, err
-		}
-		s.pool.setPool(nodeID, p)
+	ni, err := s.MetaStore.Node(nodeID)
+	if err != nil {
+		return nil, err
 	}
-	return s.pool.conn(nodeID)
-}
+	conn, err := net.Dial("tcp", ni.Host)
+	if err != nil {
+		return nil, err
+	}
 
-type remoteShardConn interface {
-	io.ReadWriter
-	Close() error
-	MarkUnusable()
+	// Write the cluster multiplexing header byte
+	conn.Write([]byte{MuxHeader})
+
+	return conn, nil
 }
 
 // RemoteMapper implements the tsdb.Mapper interface. It connects to a remote node,
@@ -93,12 +85,12 @@ type RemoteMapper struct {
 	tagsets []string
 	fields  []string
 
-	conn             remoteShardConn
+	conn             net.Conn
 	bufferedResponse *MapShardResponse
 }
 
 // NewRemoteMapper returns a new remote mapper using the given connection.
-func NewRemoteMapper(c remoteShardConn, shardID uint64, stmt influxql.Statement, chunkSize int) *RemoteMapper {
+func NewRemoteMapper(c net.Conn, shardID uint64, stmt influxql.Statement, chunkSize int) *RemoteMapper {
 	return &RemoteMapper{
 		conn:      c,
 		shardID:   shardID,
@@ -128,14 +120,12 @@ func (r *RemoteMapper) Open() (err error) {
 
 	// Write request.
 	if err := WriteTLV(r.conn, mapShardRequestMessage, buf); err != nil {
-		r.conn.MarkUnusable()
 		return err
 	}
 
 	// Read the response.
 	_, buf, err = ReadTLV(r.conn)
 	if err != nil {
-		r.conn.MarkUnusable()
 		return err
 	}
 
@@ -180,7 +170,6 @@ func (r *RemoteMapper) NextChunk() (chunk interface{}, err error) {
 		// Read the response.
 		_, buf, err := ReadTLV(r.conn)
 		if err != nil {
-			r.conn.MarkUnusable()
 			return nil, err
 		}
 

--- a/cluster/shard_mapper_test.go
+++ b/cluster/shard_mapper_test.go
@@ -5,6 +5,7 @@ import (
 	"encoding/json"
 	"fmt"
 	"io"
+	"net"
 	"testing"
 
 	"github.com/influxdb/influxdb/influxql"
@@ -13,6 +14,7 @@ import (
 
 // remoteShardResponder implements the remoteShardConn interface.
 type remoteShardResponder struct {
+	net.Conn
 	t       *testing.T
 	rxBytes []byte
 
@@ -41,8 +43,7 @@ func newRemoteShardResponder(outputs []*tsdb.MapperOutput, tagsets []string) *re
 	return r
 }
 
-func (r remoteShardResponder) MarkUnusable() { return }
-func (r remoteShardResponder) Close() error  { return nil }
+func (r remoteShardResponder) Close() error { return nil }
 func (r remoteShardResponder) Read(p []byte) (n int, err error) {
 	return io.ReadFull(r.buffer, p)
 }

--- a/tsdb/mapper.go
+++ b/tsdb/mapper.go
@@ -88,6 +88,12 @@ func (lm *SelectMapper) Open() error {
 		return lm.remote.Open()
 	}
 
+	// This can happen when a shard has been assigned to this node but we have not
+	// written to it so it may not exist yet.
+	if lm.shard == nil {
+		return nil
+	}
+
 	var err error
 
 	// Get a read-only transaction.

--- a/tsdb/show_measurements.go
+++ b/tsdb/show_measurements.go
@@ -29,9 +29,6 @@ func (e *ShowMeasurementsExecutor) Execute() <-chan *influxql.Row {
 	// Create output channel and stream data in a separate goroutine.
 	out := make(chan *influxql.Row, 0)
 
-	// It's important that all resources are released when execution completes.
-	defer e.close()
-
 	go func() {
 		// Open the mappers.
 		for _, m := range e.mappers {
@@ -104,6 +101,8 @@ func (e *ShowMeasurementsExecutor) Execute() <-chan *influxql.Row {
 		}
 
 		close(out)
+		// It's important that all resources are released when execution completes.
+		e.close()
 	}()
 	return out
 }

--- a/tsdb/show_measurements.go
+++ b/tsdb/show_measurements.go
@@ -144,18 +144,20 @@ func (m *ShowMeasurementsMapper) Open() error {
 
 	var measurements Measurements
 
-	// If a WHERE clause was specified, filter the measurements.
-	if m.stmt.Condition != nil {
-		var err error
-		measurements, err = m.shard.index.measurementsByExpr(m.stmt.Condition)
-		if err != nil {
-			return err
+	if m.shard != nil {
+		// If a WHERE clause was specified, filter the measurements.
+		if m.stmt.Condition != nil {
+			var err error
+			measurements, err = m.shard.index.measurementsByExpr(m.stmt.Condition)
+			if err != nil {
+				return err
+			}
+		} else {
+			// Otherwise, get all measurements from the database.
+			measurements = m.shard.index.Measurements()
 		}
-	} else {
-		// Otherwise, get all measurements from the database.
-		measurements = m.shard.index.Measurements()
+		sort.Sort(measurements)
 	}
-	sort.Sort(measurements)
 
 	// Create a channel to send measurement names on.
 	ch := make(chan string)

--- a/tsdb/store.go
+++ b/tsdb/store.go
@@ -305,10 +305,6 @@ func (s *Store) WriteToShard(shardID uint64, points []Point) error {
 
 func (s *Store) CreateMapper(shardID uint64, stmt influxql.Statement, chunkSize int) (Mapper, error) {
 	shard := s.Shard(shardID)
-	if shard == nil {
-		// This can happen if the shard has been assigned, but hasn't actually been created yet.
-		return nil, nil
-	}
 
 	switch st := stmt.(type) {
 	case *influxql.SelectStatement:

